### ISSUE TITLE
Fuzzer is broken

### ISSUE THREAD

**Issue by zormit** [ethersync/ethersync#125]
> I can also reproduce the failing fuzzer locally sometimes (1/5 runs or so?).
>
> Ideas for how to proceed:
>
> - [ ] Enable debugging in the fuzzer, and look hard at the output?
> - [ ] Log errors in the Vim plugin, and see if there's a failing JSON-RPC request there?
> - [ ] Check if it's always the case that daemon == nvim and peer == nvim2 (is it a peering problem?)
> - [x] Change the fuzzer so that it runs relatively fast locally (decrease sleeps), but increase number of steps, and with that configuration, bisect and see where it starts failing. Seemed to be around the new "open" behaviour, right?
> - [ ] Consider that's is a real, deep issue that leads to incorrect results. It's just surprising that we started seeing this a couple of days ago, and the fuzzer seemed fine before that.

_Originally posted by @blinry in https://github.com/ethersync/ethersync/issues/109#issuecomment-2284792055_

**Comment by zormit** [comment]
I tried bisecting. I took on the 4th step "Change the fuzzer so that it runs relatively fast locally... increase number of steps, and with that configuration, bisect and see where it starts failing"

- No timing changes apart from what's in the repo at the time
- Increased number of random edits to 500. That failed quite reliably in the latest commits, while 100 didn't show failure so easily.
- Using release build was fast enough for my taste, I did some vocabulary while waiting on some failure timeouts 🤓 (successes don't take so long)

When I had at least three successes in a row, I marked it as good. When I had a failure before reaching that, I marked it as bad. 

```bash
% git bisect log
git bisect start
# status: waiting for both good and bad commits
# bad: [065af4a9c9ca189203f6d8959f637ba7e9debf91] Merge pull request #124 from ethersync/faster-ci
git bisect bad 065af4a9c9ca189203f6d8959f637ba7e9debf91
# status: waiting for good commit(s), bad commit known
# bad: [1c4e1a20ac9939a3e8ac79e0591776526c6de8f7] Merge pull request #98 from Sohalt/main
git bisect bad 1c4e1a20ac9939a3e8ac79e0591776526c6de8f7
# status: waiting for good commit(s), bad commit known
# good: [c8521cc04b7995a80dfc732cf84cd92164b0f77b] Merge pull request #95 from ethersync/revert-nix-improvement
git bisect good c8521cc04b7995a80dfc732cf84cd92164b0f77b
# good: [24a0fb8c86b228841655355a50847c812f169fe5] Fix fuzzer: Don't require a Peer ID to connect
git bisect good 24a0fb8c86b228841655355a50847c812f169fe5
# bad: [84902de96c8e129d9391a77f1e800d929dfb17dc] Read configuration from config file in .ethersync
git bisect bad 84902de96c8e129d9391a77f1e800d929dfb17dc
# good: [755f398e79a86bf44f4fd095b7a98cb505e62d45] Use a shared secret create a libp2p "Private Network"
git bisect good 755f398e79a86bf44f4fd095b7a98cb505e62d45
# bad: [d57a9f0d727360464ae9d2c26696afcb41456aa6] When our sync protocol handler fails, shut down SyncActor, as well
git bisect bad d57a9f0d727360464ae9d2c26696afcb41456aa6
# first bad commit: [d57a9f0d727360464ae9d2c26696afcb41456aa6] When our sync protocol handler fails, shut down SyncActor, as well
```

**Comment by blinry** [comment]
Confirming: Looks like it's d57a9f0d727360464ae9d2c26696afcb41456aa6! :O In the previous commit, fuzzing with 500 steps worked for me 5 times, as well, while it failed for the faulty commit 3 times.

Commenting out the select! branch for `self.shutdown_token.cancelled()` makes the fuzzer run again.

Interestingly, by logging, I couldn't detect the cancellation tokens every being used yet.

**Comment by blinry** [comment]
My attempted fix in #126 is still flaky…

I tried [replicating a .cancelled loop in the Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=use+tokio_util%3A%3Async%3A%3ACancellationToken%3B%0Ause+tokio%3A%3A%7Bspawn%2C+select%7D%3B%0Ause+tokio%3A%3Atime%3A%3Asleep%3B%0Ause+std%3A%3Atime%3A%3ADuration%3B%0A%0A%23%5Btokio%3A%3Amain%5D%0Aasync+fn+main%28%29+%7B%0A++++let+token+%3D+CancellationToken%3A%3Anew%28%29%3B%0A++++let+cloned_token+%3D+token.clone%28%29%3B%0A++++%0A++++spawn%28async+move+%7B%0A+++++++loop+%7B%0A+++++++++++select%21+%7B%0A+++++++++++++++_+%3D+cloned_token.cancelled%28%29+%3D%3E+%7B%0A+++++++++++++++++++println%21%28%22cancelled%22%29%3B%0A+++++++++++++++%7D%0A+++++++++++++++_+%3D+sleep%28Duration%3A%3Afrom_millis%28100%29%29+%3D%3E+%7B%0A+++++++++++++++++++println%21%28%22tick%22%29%3B%0A+++++++++++++++%7D%0A+++++++++++%7D%0A+++++++%7D+%0A++++%7D%29%3B%0A++++%0A++++sleep%28Duration%3A%3Afrom_secs%281%29%29.await%3B%0A++++token.cancel%28%29%3B%0A%7D), to see if there's a problem with it, but that seems perfectly fine.

**Comment by blinry** [comment]
Maybe it was flaky because with 500 iterations in the fuzzer, it seems to need 1:40 minutes to converge. Out of caution, I increased the timeout to 5 minutes.

I *think* this seems fine now. #109, rebased on #126, also is green.
